File: app/src/main/java/id/kuato/woahelper/main/Button.java
package id.kuato.woahelper.main;

import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.view.LayoutInflater;

import androidx.annotation.Nullable;
import androidx.annotation.StringRes;

import com.google.android.material.card.MaterialCardView;

import id.kuato.woahelper.R;
import id.kuato.woahelper.databinding.ButtonBinding;

public class Button extends MaterialCardView {
    private final ButtonBinding layout;
    private final Context context;

    public Button(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        layout = ButtonBinding.inflate(LayoutInflater.from(context), this, true);
        this.context = context;
        TypedArray a = context.getTheme().obtainStyledAttributes(attrs, R.styleable.Button, 0, 0);
        if (attrs == null) return;
        String title = a.getString(R.styleable.Button_title), subtitle = a.getString(R.styleable.Button_subtitle);
        int imageId = a.getResourceId(R.styleable.Button_image, R.drawable.ic_disk);
        layout.title.setText(title);
        layout.subtitle.setText(subtitle);
        layout.image.setImageResource(imageId);
    }

    public void setTitle(@StringRes int stringId) {
        layout.title.setText(context.getString(stringId));
    }

    public void setTitle(String title) {
        layout.title.setText(title);
    }

    public void setSubtitle(@StringRes int stringId) {
        layout.subtitle.setText(context.getString(stringId));
    }

    public void setSubtitle(String subtitle) {
        layout.subtitle.setText(subtitle);
    }
}

File: app/src/main/kotlin/com/woa/helper/main/Button.kt
package com.woa.helper.main

import android.content.Context
import android.graphics.drawable.GradientDrawable
import android.util.AttributeSet
import android.view.LayoutInflater
import android.widget.FrameLayout
import com.woa.helper.R
import com.woa.helper.databinding.ButtonBinding

class Button(context: Context, attrs: AttributeSet?) : FrameLayout(context, attrs) {
    private val layout: ButtonBinding = ButtonBinding.inflate(LayoutInflater.from(context), this, true)

    private val cornerRadiusPx = resources.getDimension(R.dimen.cardCornerRadius)
    private val strokeWidthPx = resources.getDimensionPixelSize(R.dimen.strokeWidth)
    private val fillColor = context.getColor(R.color.md_theme_inverseOnSurface)
    private val strokeColor = context.getColor(R.color.md_theme_outlineVariant)

    init {
        val a = context.theme.obtainStyledAttributes(attrs, R.styleable.Button, 0, 0)
        if (attrs != null) {
            val title = a.getString(R.styleable.Button_title)
            val subtitle = a.getString(R.styleable.Button_subtitle)
            val imageId = a.getResourceId(R.styleable.Button_image, R.drawable.ic_disk)
            layout.title.text = title
            layout.subtitle.text = subtitle
            layout.image.setImageResource(imageId)
        }

        clipToOutline = true
        updateBackground()
    }

    override fun setEnabled(enabled: Boolean) {
        super.setEnabled(enabled)
        updateBackground()
    }

    private fun updateBackground() {
        alpha = if (isEnabled) 1f else 0.5f
        background = GradientDrawable().apply {
            shape = GradientDrawable.RECTANGLE
            cornerRadius = cornerRadiusPx
            setColor(fillColor)
            setStroke(if (isEnabled) strokeWidthPx else 0, strokeColor)
        }
    }

    fun setTitle(stringId: Int) {
        layout.title.text = context.getString(stringId)
    }

    fun setTitle(title: String?) {
        layout.title.text = title
    }

    fun setSubtitle(stringId: Int) {
        layout.subtitle.text = context.getString(stringId)
    }

    fun setSubtitle(subtitle: String?) {
        layout.subtitle.text = subtitle
    }
}
